compiler/rustc_trait_selection/src/solve/eval_ctxt.rs

```diff
@@ -48,7 +48,15 @@ pub(super) enum IsNormalizesToHack {
 
 #[derive(Debug, Clone)]
 pub(super) struct NestedGoals<'tcx> {
+    /// This is a normalizes-to goal that is treated specially during the evaluation
+    /// loop. We solve it by taking the RHS of the projection, replacing it with
+    /// a fresh inference variable, and only after response we equate it with the
+    /// RHS of the predicate.
+    ///
+    /// This is both to improve caching, and to avoid using the RHS of the
+    /// projection predicate to influence the normalizes-to candidate we select.
```

Comment:
For me the important (and subtle) point here is something like that: 

```suggestion
    /// This normalizes-to goal that is treated specially during the evaluation
    /// loop. In each iteration we take the RHS of the projection, replace it with
    /// a fresh inference variable, and only after evaluating that goal do we
    /// equate the fresh inference variable with the actual RHS of the predicate.
    ///
    /// This is both to improve caching, and to avoid using the RHS of the
    /// projection predicate to influence the normalizes-to candidate we select.
    ///
    /// This is not a 'real' nested goal. We must end up with a solver cycle by
    /// forgetting to replace the RHS each time we evaluate this goal. With more
    /// powerful coinduction in the future that would be unsound. 
```
(wording nits in the first paragraph and an additional 3rd paragraph... thx github diff view)